File: src/main/java/requestreplyapi/requestreply/Requestor.java

package requestreplyapi.requestreply;

import requestreplyapi.common.IAddress;

import java.net.*;
import java.io.*;


public class Requestor {

    private Socket s;
    private OutputStream oStr;
    private InputStream iStr;
    private String myName;

    public Requestor(String theName) {
        myName = theName;
    }


    public byte[] deliver_and_wait_feedback(IAddress theDest, byte[] data) {

        byte[] buffer = null;
        int val;
        try {
            s = new Socket(theDest.dest(), theDest.port());
            System.out.println("Requestor: Socket" + s);
            oStr = s.getOutputStream();
            oStr.write(data.length);
            oStr.write(data);
            oStr.flush();
            iStr = s.getInputStream();
            val = iStr.read();
            buffer = new byte[val];
            iStr.read(buffer);
            iStr.close();
            oStr.close();
            s.close();
        } catch (IOException e) {
            System.out.println("IOException in deliver_and_wait_feedback");
        }
        return buffer;
    }

}

File: ObjectRequestBrokerCS/ObjectRequestBrokerCS/requestreplyapi/requestreply/Requestor.cs
using System;

namespace requestreplyapi.requestreply
{
    using ObjectRequestBrokerCS.tests;
    using System.Net.Sockets;
    using IAddress = common.IAddress;



    public class Requestor
    {

        private TcpClient client;
        private NetworkStream stream;
        private System.IO.Stream iStr;
        private string myName;

        public Requestor(string theName)
        {
            myName = theName;
        }


        public virtual byte[] deliver_and_wait_feedback(IAddress theDest, byte[] data)
        {

            byte[] buffer = null;
            int val;
            try
            {
                client = new TcpClient(theDest.dest(), theDest.port());
                Log.log("Requestor: Socket" + client.ToString());
                stream = client.GetStream();
                stream.WriteByte((byte)(data.Length >> 8));
                stream.WriteByte((byte)(data.Length & 0xFF));
                stream.Write(data, 0, data.Length);
                stream.Flush();
                val = stream.ReadByte() << 8;
                val |= stream.ReadByte();
                buffer = new byte[val];
                stream.Read(buffer, 0, buffer.Length);
            }
            catch (SocketException)
            {
                Log.log("IOException in deliver_and_wait_feedback");
            }
            finally
            {

                stream.Close();
                client.Close();
            }
            return buffer;
        }

    }


}